### PR TITLE
feat(addon-doc): support Promise content

### PR DESCRIPTION
**Pull request by splincode**
## PR Checklist

Please check if your PR fulfills the following requirements:

-   [x] The commit message follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/)
-   [ ] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

-   [ ] Bugfix
-   [x] Feature
-   [ ] Refactoring (no functional changes, no api changes)
-   [ ] Other... Please describe:

## What is the current behavior?

```ts
import {default as example1Html} from '!!raw-loader!./examples/1/index.html';
import {default as example1Ts} from '!!raw-loader!./examples/1/index.ts';

import {default as example2Html} from '!!raw-loader!./examples/2/index.html';
import {default as example2Ts} from '!!raw-loader!./examples/2/index.ts';

import {default as exampleImportModule} from '!!raw-loader!./examples/import/import-module.txt';
import {default as exampleInsertTemplate} from '!!raw-loader!./examples/import/insert-template.txt';

//
export class ExampleTuiAvatarComponent {
    readonly exampleImportModule = exampleImportModule;
    readonly exampleInsertTemplate = exampleInsertTemplate;

    readonly example1: FrontEndExample = {
        TypeScript: example1Ts,
        HTML: example1Html,
    };

    readonly example2: FrontEndExample = {
        TypeScript: example2Ts,
        HTML: example2Html,
    };
}
```

in the current codebase you need to come up with names for the variables for the source files

## What is the new behavior?

```ts
export class ExampleTuiAvatarComponent {
    readonly exampleImportModule = import('!!raw-loader!./examples/import/import-module.txt');
    readonly exampleInsertTemplate = import('!!raw-loader!./examples/import/insert-template.txt');

    readonly example1: TuiDocExample = {
        TypeScript: import('!!raw-loader!./examples/1/index.ts'),
        HTML: import('!!raw-loader!./examples/1/index.html'),
    };

    readonly example2: TuiDocExample = {
        TypeScript: import('!!raw-loader!./examples/2/index.ts'),
        HTML: import('!!raw-loader!./examples/2/index.html'),
    };
```

you can assign imports directly

## Does this PR introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

**Comment by splincode** [comment]
@waterplea @nsbarsukov @vladimirpotekhin please, review me)

File: projects/addon-doc/src/interfaces/page.ts [hunk 1]
```diff
@@ -11,3 +11,23 @@ export interface TuiDocPage extends TuiDocPageBase {
...
+
+export type TuiDocCodeValue = RawLoaderContent | string | null | undefined;
+
+export interface TuiDocExample extends Record<string, TuiDocCodeValue> {
+    [EXAMPLE_PRIMARY_FILE_NAME.TS]?: TuiDocCodeValue;
+    [EXAMPLE_PRIMARY_FILE_NAME.HTML]?: TuiDocCodeValue;
+    [EXAMPLE_PRIMARY_FILE_NAME.LESS]?: TuiDocCodeValue;
+}
+
+export type TuiDocCodeProcessedValue = string | null | undefined;
```
Comment (nsbarsukov) [inline projects/addon-doc/src/interfaces/page.ts:31]: why do we need so many `falsy` values ?
it produces a lot of optional chaining in code :(

Maybe it better to keep only `string` (and return empty string where it is necessary to indicate `falsy` value) ?

File: projects/demo/src/modules/app/classes/ts-file.parser.ts [hunk 2]
```diff
@@ -1,53 +1,61 @@
...
     }
 
     get hasNgComponent(): boolean {
-        return this.rawFileContent.includes('@Component');
+        return this.rawFileContent?.includes('@Component') ?? false;
     }
 
-    constructor(protected rawFileContent: string) {
-        const classesInside = rawFileContent.match(/export class/gi) || [];
+    constructor(protected rawFileContent: TuiDocCodeProcessedValue) {
```
Comment (nsbarsukov) [inline projects/demo/src/modules/app/classes/ts-file.parser.ts:26]: Let's turn `rawFileContent` into the string, by default

or dont create these class if `TuiDocCodeProcessedValue` is not string

Comment (nsbarsukov) [reply projects/demo/src/modules/app/classes/ts-file.parser.ts:26]: So much optional chaining (in this class and its descendants) scares me :(

File: projects/demo/src/modules/components/avatar/avatar.component.ts [hunk 3]
```diff
@@ -1,34 +1,30 @@
...
+    readonly exampleInsertTemplate = import(
+        '!!raw-loader!./examples/import/insert-template.txt'
+    );
 
-    readonly example1: FrontEndExample = {
-        TypeScript: example1Ts,
-        HTML: example1Html,
+    readonly example1: TuiDocExample = {
+        TypeScript: import('!!raw-loader!./examples/1/index.ts'),
+        HTML: import('!!raw-loader!./examples/1/index.html'),
```
Comment (nsbarsukov) [inline projects/demo/src/modules/components/avatar/avatar.component.ts:22]: cool 🔥

**Review by nsbarsukov — COMMENTED** [review]
take attention to failed screenshot test

File: projects/addon-doc/src/utils/ensure-processed-content.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+} from '../interfaces/page';
+
+export async function ensureProcessedContent(
+    content: TuiDocExample,
+): Promise<TuiDocExampleProcessed> {
+    const processedContent: TuiDocExampleProcessed = {};
+
+    for (const [key, value] of Object.entries(content)) {
+        if (value instanceof Promise) {
+            processedContent[key] = (await (value as RawLoaderContent)).default;
```
[the remark this answers is not in the document]
Comment (splincode) [reply projects/addon-doc/src/utils/ensure-processed-content.ts:14]: good catch!

File: projects/addon-doc/src/components/example/example-get-tabs.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import {Pipe, PipeTransform} from '@angular/core';
+import {TuiDocExampleProcessed} from '@taiga-ui/addon-doc';
+
+@Pipe({name: 'tuiDocExampleGetTabs'})
+export class TuiDocExampleGetTabsPipe implements PipeTransform {
+    public transform(content: TuiDocExampleProcessed, defaultTab: string): string[] {
```
Comment (waterplea) [inline projects/addon-doc/src/components/example/example-get-tabs.pipe.ts:6]: We do not write `public`. Also let's just spread the `keys`, no need for an extra `const`. It's pretty clear without it, I think.

**Comment by splincode** [comment]
@waterplea updated)

**Review by waterplea — APPROVED** [review]

**Review by nsbarsukov — APPROVED** [review]

**Pull request merged**